### PR TITLE
cnf-tests: Rewrite the cnf-tests container to contain only the latency tests

### PR DESCRIPTION
**Pull request by jlojosnegros**
The cnf-tests split implementation #1501  breaks the cnf-tests container.
We don't build the test binaries anymore nor run the functional tests in a containerized way.

cnf-tests container is a supported product for running latency tests, 
so we need to rewrite the cnf-tests container Dockerfiles to include only the latency test.

Additionally it would be a god time to improve user interface to allow users 
to easily invoke the tests without providing many different parameters and environment variables, 
thus providing better UX and preventing errors.

### CODE REVIEW

**Comment by liornoy** [comment]
Thanks @jlojosnegros. the docker build works on my end.
I'm not qualified to indicate if it contains all the tests required for the latency tests product.
I'll let a compute team member verify it.
Are you planning to work on how a client invokes this test next?

**Comment by jlojosnegros** [comment]
> Thanks @jlojosnegros. the docker build works on my end. I'm not qualified to indicate if it contains all the tests required for the latency tests product. I'll let a compute team member verify it. Are you planning to work on how a client invokes this test next?

About the latency tests, maybe @mrniranjan or @yanirq could take a look and sees if those are all the tests that need to be run.

About the invocation ... I have changed both `build-test-bin.sh` and `entrypoint/test-run.sh` and now the user should be able to use the image just specifying  the KUBECONFIG 
PTAL and tell me WDYT.

File: cnf-tests/entrypoint/test-run.sh [hunk 1]
```diff
@@ -2,20 +2,6 @@
...
-for suite in "${suites[@]}"; do
-    if [ "$DISCOVERY_MODE" == "true" ] &&  [ "$suite" == "configsuite" ]; then
-        echo "Discovery mode enabled, skipping setup"
-        continue
-    fi
-    echo running "$SUITES_PATH/$suite" "$@"
-    "$SUITES_PATH/$suite" "$@"
-done
+pushd /usr/bin/latency_testing/one/two/three/four/five/
+LATENCY_TEST_RUN=true DISCOVERY_MODE=true FEATURES=performance ./5_latency_testing.test --ginkgo.no-color --ginkgo.vv --ginkgo.flake-attempts 2 -test.failfast --ginkgo.focus "\[performance\]\ Latency\ Test" --ginkgo.junit-report ~/tmp/artifacts
```
Comment (liornoy) [inline cnf-tests/entrypoint/test-run.sh:6]: Don't we want to invoke `latency-e2e.test` as well?

Comment (liornoy) [reply cnf-tests/entrypoint/test-run.sh:6]: In the origin `test-run.sh` we had: `"$SUITES_PATH/$suite" "$@"` which passed to the test binary
all the parameters we invoked the `test-run.sh` with.
This allowed us to pass the --junit and --report flags and chose where to output the reports.
Here you hardwire only the default `ginkgo.junit-report` to ~/tmp/artifacts. 
Should we keep the `$@` so that the user could specify values for the junit and reporter flags?

Comment (liornoy) [reply cnf-tests/entrypoint/test-run.sh:6]: Not only that but after this change, clients should be able to do the same as before, by the [documentation](https://docs.openshift.com/container-platform/4.13/scalability_and_performance/cnf-performing-platform-verification-latency-tests.html):
1. Running hwlatdetect
2. Running cyclictest
3. Running oslat
4. Generating a latency test failure report
5. Generating a JUnit latency test report
6. Running latency tests on an SNO
7. Running latency tests in a disconnected cluster
 
 So I think the `$@` is necessary for more than just the reports.
 And regarding all the requirements (1-7) we should remember to test
 if each one is working as usual after this change.

Comment (jlojosnegros) [reply cnf-tests/entrypoint/test-run.sh:6]: After a talk with @shajmakh I have found I was approaching this the wrong way, so forget about my answer.
I need to rework the PR

Comment (liornoy) [reply cnf-tests/entrypoint/test-run.sh:6]: ok, sounds good! thanks for the update.

Comment (jlojosnegros) [reply cnf-tests/entrypoint/test-run.sh:6]: > 

yeah that would not be a problem

**Comment by liornoy** [comment]
lgtm thanks, I'll wait for another approval from someone else and I'll cherry-pick it to the cnf-tests split PR

**Comment by ffromani** [comment]
if this is ready for review please remove the `[wip]` tag

**Comment by ffromani** [comment]
are the build failures expected? (e.g. `images` and `ci` lanes)

**Comment by jlojosnegros** [comment]
> are the build failures expected? (e.g. `images` and `ci` lanes)

Yes they are expected. 
This change is expecting a totally different folder layout for the tests based on git submodules.
it should work as part of  @liornoy change #1501 
https://github.com/openshift-kni/cnf-features-deploy/pull/1510#issuecomment-1558628631

File: cnf-tests/entrypoint/test-run.sh [hunk 2]
```diff
@@ -3,19 +3,19 @@ set -e
 # Setting -e is fine as we want both config and validiation to succeed
 # before running the "real" tests.
 
-TEST_SUITES=${TEST_SUITES:-"validationsuite configsuite cnftests"}
+LATENCY_TEST_RUN="${LATENCY_TEST_RUN:true}"
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:6]: Did you try to set a default value here?
Because it should be `"${LATENCY_TEST_RUN:=true}"` note the `=`

File: cnf-tests/entrypoint/test-run.sh [hunk 3]
```diff
@@ -3,19 +3,19 @@ set -e
...
+DISCOVERY_MODE="${DISCOVERY_MODE:true}"
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:7]: dito

File: cnf-tests/entrypoint/test-run.sh [hunk 4]
```diff
@@ -3,19 +3,19 @@ set -e
...
+FEATURES="${FEATURES:performance}"
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:8]: dito

File: cnf-tests/hack/build-test-bin.sh [hunk 5]
```diff
@@ -16,27 +16,23 @@ export GOFLAGS="${GOFLAGS:-"-mod=vendor"}"
...
-
-build_and_move_suite "e2esuite" "./bin/cnftests"
-build_and_move_suite "configsuite" "./bin/configsuite"
-build_and_move_suite "validationsuite" "./bin/validationsuite"
-
 if [ "$DONT_REBUILD_TEST_BINS" == "false" ] || [ -f ./cnf-tests/bin/mirror ]; then
+  echo "Building mirror ... "
   go build -o ./bin/mirror mirror/mirror.go
+  retVal=$?
+  if [ $retVal -ne 0 ]; then
```
Comment (Tal-or) [inline cnf-tests/hack/build-test-bin.sh:23]: those checks aren't necessary because you have `set -e` at the beginning of the file.

File: cnf-tests/hack/build-test-bin.sh [hunk 6]
```diff
@@ -16,27 +16,23 @@ export GOFLAGS="${GOFLAGS:-"-mod=vendor"}"
...
+    echo "Error while building mirror: ${retVal}"
+    exit $retVal
+  fi
+fi
+
+pushd submodules/cluster-node-tuning-operator
+echo "Building latency tools ... "
+make dist-latency-tests
+retVal=$?
+if [ $retVal -ne 0 ]; then
```
Comment (Tal-or) [inline cnf-tests/hack/build-test-bin.sh:33]: dito

File: cnf-tests/entrypoint/test-run.sh [hunk 7]
```diff
@@ -3,19 +3,19 @@ set -e
...
+TEST_SUITES=${TEST_SUITES:-"latency-e2e.test"}
 SUITES_PATH="${SUITES_PATH:-~/usr/bin}"
 
-suites=( $TEST_SUITES )
+suites=($TEST_SUITES)
 if [ "$IMAGE_REGISTRY" != "" ] && [[ "$IMAGE_REGISTRY" != */ ]]; then
     export IMAGE_REGISTRY="$IMAGE_REGISTRY/"
 fi
 
 for suite in "${suites[@]}"; do
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:18]: IIUC now we only have the `latency-e2e.test`, so why do we need this loop?

**Review by Tal-or — COMMENTED** [review]
Thank you for the PR

Please add the PR description as part of the commit.
This would make it easier to track and understand the change

File: cnf-tests/entrypoint/test-run.sh (hunk 7 above)
Comment (jlojosnegros) [reply cnf-tests/entrypoint/test-run.sh:18]: Because the intention is for the script to be able to run different test-suites if the user wants to.

File: cnf-tests/entrypoint/test-run.sh (hunk 2 above)
Comment (jlojosnegros) [reply cnf-tests/entrypoint/test-run.sh:6]: Not trying to assign a default value but to get the assigned value or the default (true here) if variable is not defined.
In any case it was wrong.
Thanks for pointing the problem. Nice catch

File: cnf-tests/hack/build-test-bin.sh (hunk 5 above)
Comment (jlojosnegros) [reply cnf-tests/hack/build-test-bin.sh:23]: you're right.
thanks for the catch.

**Comment by jlojosnegros** [comment]
> @jlojosnegros do we need to also rewrite https://github.com/openshift-kni/cnf-features-deploy/blob/master/hack/run-functests.sh latency parts or this will be done in #1501 alone ?

The target for this PR it is just to provide the image for cnf-tests, so we do not need to rewrite anything else here.
AFAIK #1501 aims to rewrite all tests to use submodules instead of vendoring the test dependencies so, I assume we do not need to rewrite them  ... but maybe @liornoy could correct me if I said something wrong here

File: cnf-tests/entrypoint/test-run.sh (hunk 7 above)
Comment (Tal-or) [reply cnf-tests/entrypoint/test-run.sh:18]: But this whole split purpose is to support a single suite case, so why not simplified as much as possible?

Comment (liornoy) [reply cnf-tests/entrypoint/test-run.sh:18]: @jlojosnegros sorry for the late reply, I was sick last week.
Tal-or's point is valid, i.e. with this split, the test-run.sh script will be used only for the latency test,
therefore we can simplify it and remove the loop, as it only needs to run the latency-e2e.test.

Comment (jlojosnegros) [reply cnf-tests/entrypoint/test-run.sh]: Do not worry at all @liornoy 
Change done. 
Thanks

File: cnf-tests/entrypoint/test-run.sh [hunk 8]
```diff
@@ -3,19 +3,13 @@ set -e
...
 
-for suite in "${suites[@]}"; do
-    if [ "$DISCOVERY_MODE" == "true" ] &&  [ "$suite" == "configsuite" ]; then
-        echo "Discovery mode enabled, skipping setup"
-        continue
-    fi
-    echo running "$SUITES_PATH/$suite" "$@"
-    "$SUITES_PATH/$suite" "$@"
-done
+echo running "/usr/bin/latency-e2e.test" "$@"
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:14]: the `$@` used to specify complete list of variables in the array, but we don't this array anymore so it can be removed as well.

File: cnf-tests/entrypoint/test-run.sh [hunk 9]
```diff
@@ -3,19 +3,13 @@ set -e
...
+LATENCY_TEST_RUN="$LATENCY_TEST_RUN" DISCOVERY_MODE="$DISCOVERY_MODE" FEATURES="$FEATURES" "/usr/bin/latency-e2e.test" "$@"
```
Comment (Tal-or) [inline cnf-tests/entrypoint/test-run.sh:15]: ditto

File: cnf-tests/Dockerfile [hunk 10]
```diff
@@ -81,15 +57,16 @@ COPY --from=builder-latency-test-runners /hwlatdetect-runner /usr/bin/hwlatdetec
...
-COPY --from=builder /go/src/github.com/openshift-kni/cnf-features-deploy/cnf-tests/bin/cnftests-sha.txt /usr/local/etc/cnf
+
+COPY --from=go-builder /app/cnf-tests/bin/mirror /usr/bin/mirror
+COPY --from=go-builder /app/cnf-tests/mirror/images.json /usr/local/etc/cnf
+COPY --from=go-builder /app/cnf-tests/entrypoint/test-run.sh /usr/bin/test-run.sh
+
 COPY tools/oot-driver/ /usr/src/oot-driver/
 
+# Copy test suite
+COPY --from=go-builder /app/cnf-tests/submodules/cluster-node-tuning-operator/build/_output/bin/latency-e2e.test /usr/bin/latency-e2e.test
```
Comment (shajmakh) [inline cnf-tests/Dockerfile:68]: why is this needed?

Comment (shajmakh) [reply cnf-tests/Dockerfile:68]: This is a binary load file of the latency tools tests. We use it in https://github.com/openshift/cluster-node-tuning-operator/tree/master/test/e2e/performanceprofile/functests/5_latency_testing which AFAIU don't need to export it in cnf-tests

Comment (shajmakh) [reply cnf-tests/Dockerfile:68]: Thanks for clarifying. the direction then looks good from my end.
